src/nvim/getchar.c

```diff
@@ -1577,729 +1577,843 @@ vungetc ( /* unget one character (can only be done once!) */
+    // If there was no mapping, use the character from the typeahead
+    // buffer right here. Otherwise, use the mapping (loop around).
+    // get a character: 2. from the typeahead buffer
+    int c = typebuf.tb_buf[typebuf.tb_off] & 255;
+    if (advance) {                  // remove chars from tb_buf
+      cmd_silent = (typebuf.tb_silent > 0);
+      if (typebuf.tb_maplen > 0) {
+        KeyTyped = false;
+      } else {
+        KeyTyped = true;
+        // write char to script file(s)
+        gotchars(typebuf.tb_buf + typebuf.tb_off, 1);
+      }
+      KeyNoremap = typebuf.tb_noremap[typebuf.tb_off];
+      del_typebuf(1, 0);
+    }
+    return c;
+  }
+
+  if (expand_matched_map(mp, *keylenp, mapdepthp)) {
+    return 0;
+  }
+
+  return -1;
+}
+
+static int ins_esc_special_case(
+    int *new_wcolp, int *new_wrowp, bool *mode_deletedp,
+    const int advance, const int keylen)
+{
+  // special case: if we get an <ESC> in insert mode and there
+  // are no more characters at once, we pretend to go out of
+  // insert mode.  This prevents the one second delay after
+  // typing an <ESC>.  If we get something after all, we may
+  // have to redisplay the mode. That the cursor is in the wrong
+  // place does not matter.
+  int bytes_read = 0;
+  if (       advance
+      && typebuf.tb_len == 1
+      && typebuf.tb_buf[typebuf.tb_off] == ESC
+      && !no_mapping
+      && ex_normal_busy == 0
+      && typebuf.tb_maplen == 0
+      && (State & INSERT)
+      && (p_timeout
+          || (keylen == KEYLEN_PART_KEY && p_ttimeout))
+      && (bytes_read = inchar(typebuf.tb_buf + typebuf.tb_off + typebuf.tb_len,
+                              3, 25L, typebuf.tb_change_cnt)) == 0) {
+    colnr_T col = 0, vcol;
+    char_u      *ptr;
+
+    if (mode_displayed) {
+      unshowmode(true);
+      *mode_deletedp = true;
+    }
+    validate_cursor();
+    int old_wcol = curwin->w_wcol;
+    int old_wrow = curwin->w_wrow;
+
+    // move cursor left, if possible
+    if (curwin->w_cursor.col != 0) {
+      if (curwin->w_wcol > 0) {
+        if (did_ai) {
+          // We are expecting to truncate the trailing
+          // white-space, so find the last non-white
+          // character -- webb
+          col = vcol = curwin->w_wcol = 0;
+          ptr = get_cursor_line_ptr();
+          while (col < curwin->w_cursor.col) {
+            if (!ascii_iswhite(ptr[col])) {
+              curwin->w_wcol = vcol;
             }
-            if (has_mbyte && col > 0 && curwin->w_wcol > 0) {
-              /* Correct when the cursor is on the right halve
-               * of a double-wide character. */
-              ptr = get_cursor_line_ptr();
-              col -= (*mb_head_off)(ptr, ptr + col);
-              if ((*mb_ptr2cells)(ptr + col) > 1)
-                --curwin->w_wcol;
+            vcol += lbr_chartabsize(ptr, ptr + col, (colnr_T)vcol);
+            if (has_mbyte) {
+              col += (*mb_ptr2len)(ptr + col);
```

Comment:
Also with larger refactorings we should change `(*mb_func)(` and `MB_FUNC(` to to `mb_func(`. The only reason against would be merging from vim easier but with heavy refactoring applying patches must be done manually anyway.